### ISSUE TITLE
Support v2ray-plugin

### ISSUE THREAD

**Issue by laventofly** [Jigsaw-Code/outline-client#551]
V2ray-plugin is a very useful plugin in order to improve the concealment of shadowsocks. This is almost necessary to use in China. Please consider this request again. Thank you for your outstanding work.

https://github.com/shadowsocks/v2ray-plugin

**Comment by 20051615** [comment]
Indeed, I use Outline on the client side for Windows and iOS, while deploying shadowsocks-libev on the servers (without using Outline manager). Would be great to see plugin support from Outline.

**Comment by hadifarnoud** [comment]
+1

**Comment by stonedreamforest** [comment]
+1 @JonathanDCohen  @trevj @alalamav  

https://github.com/v2ray/v2ray-core

**Comment by JonathanDCohen** [comment]
Hey there!  Sorry if I'm being dense (just joined the team and am figuring things out on the go) but I don't see anywhere what this would actually do for us.  It looks like v2ray is a platform for implementing shadowsocks to me.

**Comment by hadifarnoud** [comment]
I thought v2ray plugin is for tunnel encryption. Please check the official shadowsocks plugin here 

https://github.com/shadowsocks/v2ray-plugin

**Comment by stonedreamforest** [comment]
> Hey there! Sorry if I'm being dense (just joined the team and am figuring things out on the go) but I don't see anywhere what this would actually do for us. It looks like v2ray is a platform for implementing shadowsocks and other proxies to me. Any links to resources?

- when we use shadowsocks/outline
**`ISP`** :  _I don't know what you are doing, so I will block you...._

- when we use v2ray
**`ISP`** :  _this guy is doing looks normal. looking for the next guy..._
 
--------------------

#### https://www.v2ray.com/en/
- Obfuscation: V2Ray has built in obfuscation to hide traffic in TLS, and can run in parallel with web servers.
>  websocket + TLS + nginx/caddy ： So isp see the normal traffic access content



- Reverse proxy: General support of reverse proxy. Can be used to build tunnels to localhost.
> This feature is very good, shadowsocks/outline can't working properly in my company, but v2ray can do this

------------------

some web page may be need use google translate 


1. https://github.com/shadowsocks/v2ray-plugin/issues/31
```
just get a free domain and host it behind cloudflare so that way they have to block cloudflare's ip instead of the servers.
```


2. https://www.reddit.com/r/dumbclub/comments/aeyprm/shadowsocksr_vs_v2ray_vs_wireguard/

3. https://steemit.com/cn/@v2ray/47wumq
4. https://steemit.com/cn/@v2ray/6knmmb
see  **` Offensive and defensive battle 3.0`**

5. http://www.lbmdok.com/1.html

**Comment by 20051615** [comment]
> Hey there! Sorry if I'm being dense (just joined the team and am figuring things out on the go) but I don't see anywhere what this would actually do for us. It looks like v2ray is a platform for implementing shadowsocks and other proxies to me. Any links to resources?

I believe most people are interested in the now-deprecated https://github.com/shadowsocks/simple-obfs plugin, which is now replaced by the more multifunctional and up-to-date v2ray-plugin. Some people would argue that the effectiveness of obfuscation is an untested myth, though.

**Comment by moralrebuild** [comment]
A must have feature, otherwise the shadowsocks already proves itself as a broken VPN solution from last month in China.

**Comment by clementhk** [comment]
Can confirm this works in Mainland China and UAE.

**Comment by joeytwiddle** [comment]
I have posted a [$200 bounty](https://www.bountysource.com/issues/71913530-support-v2ray-plugin) on this issue, for anyone who can get this working in the next 3 months. 😉

(For the bounty I would accept just Android and iOS support, I don't care about desktop support.)

**Comment by clementhk** [comment]
> I have posted a [$200 bounty](https://www.bountysource.com/issues/71913530-support-v2ray-plugin) on this issue, for anyone who can get this working in the next 3 months. 😉
> 
> (For the bounty I would accept just Android and iOS support, I don't care about desktop support.)

I'll follow up with another $200 if anyone is working on it.
And I do care about desktop support.
But since the client uses shadowsocks-libev which supports v2ray-plugin already, it should be pretty easy.

**Comment by cjhenck** [comment]
What specific elements of V2Ray are people interested in having in Outline? It seems like most people are advocating for WebSockets + proxying over TLS, which could appear to be "web traffic" from a network perspective.

There's also VMess, which is V2Ray's proprietary protocol. Is that a useful addition?

**Comment by clementhk** [comment]
No need VMess for me.

**Comment by Keung123** [comment]
Since last year, Chinese firewall firgure out how to recognise and block SS traffic. SS is dead for mainland Chinese users. So it would be great to see outline, such a supportive vpn solution, to have vmess support.

**Comment by Keung123** [comment]
> What specific elements of V2Ray are people interested in having in Outline? It seems like most people are advocating for WebSockets + proxying over TLS, which could appear to be "web traffic" from a network perspective.
> 
> There's also VMess, which is V2Ray's proprietary protocol. Is that a useful addition?

Since last year, Chinese firewall firgure out how to recognise and block SS traffic. SS is basically dead for mainland Chinese users, a new-setup SS server could be quickly block. So it would be great to see outline, such a supportive vpn solution, to have vmess support.

**Comment by clementhk** [comment]
@Keung123, Shadowsocks + V2Ray works in China. Adding another protocol in Outline is a bigger job so please open another issue instead of hijacking this issue, thanks. Adding a Shadowsocks plugin is much smaller job than adding vmess as now the client already uses shadowsocks-libev.
Our office in China uses Shadowsocks + V2Ray and it has been working well.

**Comment by Keung123** [comment]
Does it configured https proxy on v2ray server? In my case, v2ray over ss didn't work. But I could not confirm if the sever had been blocked already before I run v2ray.

**Comment by hadifarnoud** [comment]
VMess works like a charm in Iran

**Comment by BGramApp** [comment]
No news about v2ray plugin in Outline?

**Comment by hadifarnoud** [comment]
just wanted to confirm I've been using my own VMESS setup and it works like a charm. it worth doing

**Comment by fortuna** [comment]
We've made several improvements to Outline detection resistance since this issue was filed. We have reports that it's working well in Iran and China.

Please try Outline again and let us know if you still have issues. We can open a new bug with fresh information if needed.

**Comment by Yeganloo** [comment]
Hi
I am in Iran and it doesn't work. 
This is the test result on my server:

Host [My Host] Port 443

Testing...

Test assumes port 80 on [My Host] refuses all requests

Your network does not block port 443

Port 443 on host [My Host] is blocked by packed dropping

Done!

**Comment by ghominejad** [comment]
> > Hey there! Sorry if I'm being dense (just joined the team and am figuring things out on the go) but I don't see anywhere what this would actually do for us. It looks like v2ray is a platform for implementing shadowsocks and other proxies to me. Any links to resources?
> 
> * when we use shadowsocks/outline
>   **`ISP`** :  _I don't know what you are doing, so I will block you...._
> * when we use v2ray
>   **`ISP`** :  _this guy is doing looks normal. looking for the next guy..._
> 
> #### https://www.v2ray.com/en/
> * Obfuscation: V2Ray has built in obfuscation to hide traffic in TLS, and can run in parallel with web servers.
> 
> > websocket + TLS + nginx/caddy ： so isp see the "normal" traffic when we access content
> 
> * Reverse proxy: General support of reverse proxy. Can be used to build tunnels to localhost.
> 
> > This feature is very good, shadowsocks/outline can't working properly in my company, but v2ray can do this
> 
> some web page may be need use google translate
> 
> 1. [Shadowsocks, v2ray, and simple-obfs shadowsocks/v2ray-plugin#31](https://github.com/shadowsocks/v2ray-plugin/issues/31)
> 
> ```
> just get a free domain and host it behind cloudflare so that way they have to block cloudflare's ip instead of the servers.
> ```
> 
> 2. https://www.reddit.com/r/dumbclub/comments/aeyprm/shadowsocksr_vs_v2ray_vs_wireguard/
> 3. https://steemit.com/cn/@v2ray/47wumq
> 4. https://steemit.com/cn/@v2ray/6knmmb
>    see  **` Offensive and defensive battle 3.0`**
> 5. http://www.lbmdok.com/1.html

They don't block IPs, they block protocols.  Currently VMESS is working in Iran.

**Comment by parsanoori** [comment]
Any news on v2ray plugin support?

**Comment by Negashev** [comment]
Any news on v2ray plugin support? 2x

**Comment by fortuna** [comment]
We are working on out [Outline SDK](https://github.com/Jigsaw-Code/outline-sdk), which will provide a library of transports and strategies that are cross platform and can be easily reused and composed.

Then we will need to refactor the Outline Client code to remove explicit dependencies on the protocol details from the platform-specific code, so we can change the protocol without touching every platform.

At that point, we should be able to configure different protocols, including TLS-based ones, which is what is really missing in Outline.

VMESS or V2Ray is not the solution. It's what protocol configuration works. It's often something over TLS (I see VMESS over Websockets a lot). But the actual proxy protocol can be SOCKS5, HTTP CONNECT, VMESS, Shadowsocks SOCKS, ... almost interchangeably

We have no plans to support v2ray plugins. They are not cross-platform and open up a large surface area for security vulnerability. They also often lack support for UDP.

**Comment by kklem0** [comment]
> We are working on out [Outline SDK](https://github.com/Jigsaw-Code/outline-sdk), which will provide a library of transports and strategies that are cross platform and can be easily reused and composed.
> 
> Then we will need to refactor the Outline Client code to remove explicit dependencies on the protocol details from the platform-specific code, so we can change the protocol without touching every platform.
> 
> At that point, we should be able to configure different protocols, including TLS-based ones, which is what is really missing in Outline.
> 
> VMESS or V2Ray is not the solution. It's what protocol configuration works. It's often something over TLS (I see VMESS over Websockets a lot). But the actual proxy protocol can be SOCKS5, HTTP CONNECT, VMESS, Shadowsocks SOCKS, ... almost interchangeably
> 
> We have no plans to support v2ray plugins. They are not cross-platform and open up a large surface area for security vulnerability. They also often lack support for UDP.

Yes!! I thought the project is dead..

**Comment by fortuna** [comment]
@kklem0 The project is not dead. In fact, we've never helped as many people as we do now!

**Comment by zachlai** [comment]
Thanks for all the efforts you guys made, and am so glad to see updated discussion here. 
I think different approaches are fine, but the fact I experienced in recent several years in some big cities in China is that SS is so easily blocked. I tried Outline many times, but as a result, within few days, the ips of my servers are blocked. 
If your aim is to make it easier for users to build and use, could you give more instructions on how to make it stronger and more difficult for the GFW to detect?
I'm still looking for the new features of this program and thanks again for the efforts you made on this project!

**Comment by qiangweihewu** [comment]
> We are working on our [Outline SDK](https://github.com/Jigsaw-Code/outline-sdk), which will provide a library of transports and strategies that are cross platform and can be easily reused and composed.
> 
> Then we will need to refactor the Outline Client code to remove explicit dependencies on the protocol details from the platform-specific code, so we can change the protocol without touching every platform.
> 
> At that point, we should be able to configure different protocols, including TLS-based ones, which is what is really missing in Outline.
> 
> VMESS or V2Ray is not the solution. It's what protocol configuration works. It's often something over TLS (I see VMESS over Websockets a lot). But the actual proxy protocol can be SOCKS5, HTTP CONNECT, VMESS, Shadowsocks SOCKS, ... almost interchangeably
> 
> We have no plans to support v2ray plugins. They are not cross-platform and open up a large surface area for security vulnerability. They also often lack support for UDP.

Thanks for all the efforts you guys made, and am so glad to see updated discussion here.
I think different approaches are fine, but the fact I experienced in recent several years in some big cities in China is that SS is so easily blocked. I tried Outline many times, but as a result, within few days, the ips of my servers were blocked so soon.
If your aim is to make it easier for users to build and use, could you give more instructions on how to make it stronger and more difficult for the GFW to detect?
I'm still looking for the new features of this program and thanks again for the efforts you made on this project!

**Comment by fortuna** [comment]
@qiangweihewu I put together some slides for service providers at https://drive.google.com/file/d/1EBFMu7O8dpzwnyibGXTg8VgMBFaltIJW/view

You may benefit from using [disguise prefix](https://www.reddit.com/r/outlinevpn/wiki/index/prefixing/) to reduce detection and [dynamic keys](https://www.reddit.com/r/outlinevpn/wiki/index/dynamic_access_keys/) to allow for service rotation.
